### ISSUE TITLE
[clang-tidy] `misc-const-correctness` false positive

### ISSUE THREAD

**Issue by tearfur** [llvm/llvm-project#133562]
https://godbolt.org/z/dbKn7dWYz

```
<source>:8:5: warning: variable 'vz' of type 'size_t' (aka 'unsigned long') can be declared 'const' [misc-const-correctness]
    8 |     size_t vz = static_cast<size_t>(v);
      |     ^
      |            const
```

If I change the problematic line to `auto vz = static_cast<size_t>(v);`, the warning goes away.

**Comment by tearfur** [comment]
P.S. Weird that clang-tidy does not warn `modernize-use-auto` here.

**Comment by firewave** [comment]
> P.S. Weird that clang-tidy does not warn `modernize-use-auto` here.

That warning is not enabled by default, so you need to specify it explicitly. If you enable all checks it is being reported:

```
<source>:8:5: warning: use auto when initializing with a cast to avoid duplicating the type name [hicpp-use-auto,modernize-use-auto]
    8 |     size_t vz = static_cast<size_t>(v);
      |     ^~~~~~
      |     auto
```

**Comment by tearfur** [comment]
Yes, of course it's me being stupid again. :)

Still, I think the main issue stands.

**Comment by tearfur** [comment]
Indeed looks like it.
